kernel/rros/factory.rs

```diff
@@ -252,7 +261,117 @@ impl RrosElement {
             pointer: 0 as *mut u8,
         })
     }
+
+    /// `remove_element_device`: Remove a element device from hash list
+    /// and unregister this device.
+    pub fn remove_element_device(&mut self) {
+        let fac = unsafe { &mut *(self.factory.locked_data().get()) };
+        // SAFETY: If self.dev is None, it will panic. But when we
+        // call this function, we make sure there is a device in element.
+        // So it is safe to unwrap.
+        let dev = self.dev.as_mut().unwrap();
+        dev.unregister();
+
+        if self.is_public() {
+            drop(self.cdev.take());
+        }
+
+        {
+            let _guard = fac.inside.hash_lock.as_ref().unwrap().lock();
+            self.hash.hash_del();
+        }
+    }
+
+    /// `is_public`: If element is public return ture
+    #[inline]
+    fn is_public(&self) -> bool {
+        self.clone_flags & RROS_CLONE_PUBLIC != 0x0
+    }
+
+    /// `__do_put_element`: Delete an element
+    pub fn __do_put_element(&mut self) {
+        extern "C" {
+            fn rust_helper_synchronize_rcu();
+        }
+        let fac = unsafe { &*(self.factory.locked_data().get()) };
+        if self.dev.is_none() {
+            fac.dispose.unwrap()(self);
+        }
+
+        /*
+         * self->minor won't be free for use until rros_destroy_element()
+         * is called from the disposal handler, so there is no risk of
+         * reusing it too early.
+         */
+        self.remove_element_device();
+
+        /*
+         * Serialize with rros_open_element().
+         */
+        unsafe {
+            rust_helper_synchronize_rcu();
+        }
+
+        fac.dispose.unwrap()(self);
```

Comment:
Is `dispose` a mandatory field? Maybe we should provide a way to let every subsystem pass its function pointer at initialization?

Reply:
It is now `Option<fn(&mut RrosElement)>`, subsystems can change the fn inside at the factory initialization.